docs/RFCS/query_cancellation.md

```diff
@@ -0,0 +1,189 @@
+|       2 | root     | 2017-06-07 16:14:37.381263+00:00 | UPDATE accounts SET balance = CASE id WHEN $1 THEN $3::INT WHEN $2 THEN $4::INT END WHERE id IN ($1, $2) | 127.0.0.1:54912 |                  | false       | executing |
+|       2 | root     | 2017-06-07 16:14:37.3805+00:00   | UPDATE accounts SET balance = CASE id WHEN $1 THEN $3::INT WHEN $2 THEN $4::INT END WHERE id IN ($1, $2) | 127.0.0.1:54928 |                  | false       | executing |
++---------+----------+----------------------------------+----------------------------------------------------------------------------------------------------------+-----------------+------------------+-------------+-----------+
+(6 rows)
+```
+
+Adding a unique query identifier to each query and exposing that in this table will be necessary
+for the DBA to be able to issue the correct CANCEL command.
+
+# Detailed design
+
+## Query identifiers
+
+Query identifiers will be of the format `<node_id>:<counter>`, where `<node_id>` is the ID
+of the gateway node (an integer), and `<counter>` is an `int64` counter on that gateway node
+that gets incremented for every query run on that node.
+
+Bringing the node ID into the query identifier removes the need to either:
+
+- Contact every node to figure out where that query lives, or
+- Have the user pass in two arguments to the `CANCEL QUERY` statement; both the node ID and
+the query ID.
+
+Using a colon as the delimiter instead of a comma or period makes it clear that the ID field
+is a string field, and not a float value.
+
+Query IDs will be presented in the output of `SHOW QUERIES`:
+
+```sql
+root@:26257/> SELECT node_id, id, query FROM [SHOW CLUSTER QUERIES];
++---------+----------+--------------------------------------------------------------------+
+| node_id | id       | query                                                              | 
++---------+----------+--------------------------------------------------------------------+
+| 1       | 1:43872  | SHOW CLUSTER QUERIES                                               |
+| 2       | 2:11     | INSERT INTO Students VALUES ('Robert'); DROP TABLE Students; --')  |
++---------+----------+--------------------------------------------------------------------+
+(2 rows)
+```
+
+Having the counter part be numeric also allows us to eventually support the Postgres cancellation
+syntax:
+
+```sql
+SELECT pg_cancel_backend(<query-id>);
+```
+
+Since postgresql has integer query IDs, for compatibility reasons this function could support taking
+just the counter part of the query ID as input instead of the full ID. The db could assume it's referring
+to a query on that same gateway node.
+
+However, while supporting `pg_cancel_backend` is part of the motivation to go with integer query IDs,
+actual implementation of `pg_cancel_backend` is not a key part of this project and can be deferred
+until deemed necessary.
```

Comment:
we could also consider using an int64 and use unique_rowid() as described here:
https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/parser/builtins.go#L871

as a first implementation of the id. CANCEL can decode the unique_rowid passed to it to send the cancellation to the correct node. We don't expect to get to 32K nodes any time soon :-)